Fatal error handling consolidation and revamp

tyrantlizard77 opened the pull request:
Please see individual commit messages and #11751 for more details.

High-level changes:

* Platform-specific fatal error handling functions moved to the core kernel
* Introduce platform independent enumeration of fatal error reason codes
* Simplified default error handling policy: simply hang the system if a fatal error occurs.
* _default_esf removed. Errors without exception stack simply pass around a NULL instead.
* Fatal errors will be sent to the logging module instead of printk() if enabled.
* Some formatting improvement to arch-specific error information
* Spurious interrupts on ARC now raise a fatal error
* Fixed incorrect error reason code for failed system calls on ARM
* ARM now dumps a partial register set on a fatal error
* NANO_ESF renamed to z_arch_esf_t
* Simplified and fixed sanitycheck detection of error conditions
* Some refactoring of the ARM exception code to not overload the reason code with internal state.
* LOG_PANIC moved to the system fatal error handler function

Commit:
kernel: consolidate error handling code

* z_NanoFatalErrorHandler() is now moved to common kernel code
  and renamed z_fatal_error(). Arches dump arch-specific info
  before calling.
* z_SysFatalErrorHandler() is now moved to common kernel code
  and renamed k_sys_fatal_error_handler(). It is now much simpler;
  the default policy is simply to lock interrupts and halt the system.
  If an implementation of this function returns, then the currently
  running thread is aborted.
* New arch-specific APIs introduced:
  - z_arch_system_halt() simply powers off or halts the system.
* We now have a standard set of fatal exception reason codes,
  namespaced under K_ERR_*
* CONFIG_SIMPLE_FATAL_ERROR_HANDLER deleted
* LOG_PANIC() calls moved to k_sys_fatal_error_handler()

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arm: don't use exc reason codes for internal state

We are standardizing to a arch-independent set of exception
reason codes, don't overload it with internal state of
the ARM fault handling code.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
Still a lot to do, but wanted to get some early feedback from @andyross @gnuless to make sure this is going in the direction we want.

gnuless commented:
@andrewboie Looks like the right direction to me ...

One thing: is it possible to get rid of the `_default_esf`, and just pass in `NULL` if there's no applicable stack frame?  I've not been able to figure out the reason for needing a dummy struct.

andrewboie commented:
> One thing: is it possible to get rid of the `_default_esf`, and just pass in `NULL` if there's no applicable stack frame? I've not been able to figure out the reason for needing a dummy struct.

Completely agree, will add this to the PR.

andyross commented on kernel/fatal.c:
```diff
@@ -0,0 +1,92 @@
+	}
+#if defined(CONFIG_STACK_CANARIES) || defined(CONFIG_STACK_SENTINEL) || \
+		defined(CONFIG_HW_STACK_PROTECTION) || \
+		defined(CONFIG_USERSPACE)
```
FWIW: for clarity, maybe define a "HAVE_ERR_STACK_CHK_FAIL" elsewhere?  Or just have a case for this (like 3-4 instructions and the static string) defined always, even on configurations where we can't actually emit such an error.

andrewboie replied:
yeah more trouble than it's worth, I just removed the ifdef.

andyross commented on kernel/fatal.c:
```diff
@@ -0,0 +1,92 @@
+}
+/* LCOV_EXCL_STOP */
+
+void z_fatal_error(unsigned int reason, const NANO_ESF *esf)
```
Should probably take the time to rename and document the "NANO_ESF" type.  Maybe "struct z_arch_excinfo" or something?

Also, IIRC most of the existing architectures pass a garbage/fallback value here (c.f. the "0xdeaddead" IP, etc...).  For situations where you can't synthesize a proper frame (most architectures are just lazy I think) we probably want a portable way to do that.

andrewboie replied:
agreed on all points, working on it.

andyross commented on kernel/fatal.c:
Seems like that #ifery is doing more harm than good, maybe just:

```c
#ifdef CONFIG_THREAD_NAME
     printk("Current thread ID = %p\n", _current);
#else 
    printk("Current thread ID = %p (%s)\n", _current, thread_name ? thread_name : "unknown");
#endif 
```

andrewboie replied:
done.

andyross commented on kernel/fatal.c:
```diff
@@ -0,0 +1,92 @@
+#include <fatal.h>
+
+/* LCOV_EXCL_START */
+FUNC_NORETURN __weak void z_arch_system_halt(unsigned int reason)
```
I'd put the arch_irq_lock() in here, not outside the call.   Masked interrupts are sort of part of the definition of a system halt, someone who calls this expecting a halt but forgets the lock is going to be surprised when their timeouts & driver I/O keep happening.

Also, needless to say this is going to work only with uniprocessor setups. On SMP systems, halting the system is more complicated.  We could do it portably on top of the existing sched_ipi() mechanism (basically calling this function out of the IPI handler, which itself already has a fallback for systems without a working IPI), which would probably be useful here if we want a robust weak fallback.

Alternatively, you could #ifndef CONFIG_SMP around the weak fallback and force the SMP systems to do this themselves.  But the scheme here leaves them with a silently broken halt.

andrewboie replied:
moved.

andyross commented on kernel/fatal.c:
```diff
@@ -0,0 +1,92 @@
+	ARG_UNUSED(esf);
+
+	printk("Halting system.\n");
+	(void)irq_lock();
```
This should be z_arch_irq_lock().  On SMP, irq_lock() is a spinlock with special semantics.  Also see note elsewhere, I think this wants to be part of the halt() implementation.

gnuless replied:
Even on SMP systems, `irq_lock()` will give us the behavior we want, no?  As soon as the other CPUs attempt to grab the global lock they'll just hang.  We could nudge the process along with an IPI but I don't see that this needs changing.

andrewboie replied:
for now I have changed to z_arch_irq_lock(), but I think @gnuless makes a good point. Leaving this open until we decide the best course of action.

andyross replied:
Sorry, missed this comment yesterday: it's true that irq_lock() will work to suspend local interrupts, so it will work for the task desired (at the cost of including the code for it which might otherwise be elided if the system is fully spinlockified).  But it won't work as a replacement for a portable halt of other CPUs.

That idea would work if irq_lock() was the universally used global lock, but it's not anymore.  The core kernel synchronization has been split into a bunch of independent spinlocks (which is a feature, obviously) and so there's no guarantee that anything is going to hit this at any point.  You'd need to be using a driver that happened to be using the legacy API.

gnuless replied:
I assume the point of the lock is that ISRs will grab it upon entry, no?

andyross reviewed (commented):
Some notes.  Frankly this is probably worth merging already just for the code removal benefit.

Commit:
kernel: abolish _default_esf

NANO_ESF parameters may now be NULL, indicating that no
exception frame is available.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
Update: some cleanups and _default_esf has been abolished.

Commit:
kernel: support log system for fatal errors

We introduce a new z_fatal_print() API and replace all
occurrences of exception handling code to use it.
This routes messages to the logging subsystem if enabled.
Otherwise, messages are sent to printk().

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arc: use z_fatal_error() for spurious IRQs

We shouldn't special-case this to just spin forever.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arm: fix exception reason code for bad syscall

ARM was reporting as a CPU exception and not a kernel
oops.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arm: dump registers on fatal exceptions

We had a function that did this, but it was dead code.
Move to fatal.c and call from z_arm_fatal_error().

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

pizi-nordic commented on kernel/fatal.c:
```diff
@@ -0,0 +1,133 @@
+{
+	va_list ap;
+
+	LOG_PANIC();
```
This switches logger into panic mode and there is no way back. However he z_fatal_print could is called for recoverable faults.

IMHO, we should call LOG_PANIC() only when we know that system will be halted and the log processing thread will not be called any more.

@nordic-krch: What is your opinion?

andrewboie replied:
> This switches logger into panic mode and there is no way back. However he z_fatal_print could is called for recoverable faults.

The problem here is that the policy decision on what to do with the fault doesn't take place until k_sys_fatal_error_handler() runs, and if I understand the logger correctly we need to put the logger in panic mode immediately before dumping any exception info.

andrewboie replied:
@pizi-nordic by the way this problem has always existed. I just moved the LOG_PANIC() call to a common spot.

andrewboie replied:
@pizi-nordic I filed a bug on this some time ago: #11745

andrewboie replied:
@nordic-krch is it safe to move the LOG_PANIC into the k_sys_fatal_error_handler() function? Is there any chance of dropped messages?

nordic-krch replied:
@andrewboie not calling LOG_PANIC() early will result in logs being deferred but they are not lost. They all will be flushed on `LOG_PANIC()` call and in panic all logs are processed synchronously. However, it's better to call `LOG_PANIC` as early as possible because there is a chance that log message pool is empty and then some messages might be dropped (depending on the mode oldest or newest). So entering panic should be moved to a place where we know that it's unrecoverable.

pizi-nordic replied:
If the configuration of the logger is sane, we could guarantee that last X log messages before call to LOG_PANIC() will be preserved.

The LOG_PANIC() does 2 things:
- Switches backbends to panic mode (so all logs will be processed in place).
- Renders all buffered logs.

Since the call, the LOG() entries are no longer buffered but goes directly to backends.

andrewboie replied:
> However, it's better to call LOG_PANIC as early as possible 

> So entering panic should be moved to a place where we know that it's unrecoverable.

This is contradictory. We don't know the exception is unrecoverable until the very end. I think what needs to happen is to provide a mechanism to un-panic the log system.
Pre-existing issue anyway that has its own bug open, don't think it should hold up this PR.

nordic-krch replied:
right, that sounded a bit contradictory, first part was suppose to be `as early as possible when it's determined that it's unrecoverable`.

I'm against un-panicing. If i assume correctly you may enter fault handling in a thread and scheduler/interrupts are still working and context may be switched. If so then switching to panic mode at that moment may hurt the application since all log calls are now blocking. In the future we may have flash backend which will write/erase from flash, those are time costly operations which when done in blocking way in low priority context will impact other higher priority threads.

Commit:
kernel: rename NANO_ESF

This is now called z_arch_esf_t, conforming to our naming
convention.

This needs to remain a typedef due to how our offset generation
header mechanism works.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
sanitycheck: simplify fault detection

Any fatal error will print "ZEPHYR FATAL ERROR" now, so
we don't have to maintain a set of strings in the
sanitycheck harness.py

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
I believe this is ready for review now.

aescolar commented on arch/posix/core/fatal.c:
Just a minor thing:
Could you replace this message with something like:
`posix_print_error_and_exit("Exiting due to fatal error\n");`
(This function does not add any extra \n)

andrewboie replied:
done.

aescolar reviewed (dismissed):
Just one minor thing from me

dbkinder commented on doc/guides/porting/arch.rst:
```diff
@@ -441,13 +441,10 @@ However, a real implementation is strongly recommended.
-
-See the current architecture implementations for examples.
+In the event of an unhandled CPU exception, the architecture code must call
+into :c:func:`z_fatal_error`. If the platform supports it, it may optionally
```
This could use an explanation of what z_fatal_error does.   The ``z_fatal_error`` API doxygen comments don't mention that it could call ``z_arch_system_halt``, but do talk about calling ``k_sys_fatal_error_handler()``. 
```
 * This function dumps out architecture-agnostic information about the error
 * and then makes a policy decision on what to do by invoking
 * k_sys_fatal_error_handler().
 *
```
How is the optional behavior configured for spinning or powering off?

andrewboie replied:
@dbkinder I'm not sure I understand your review.
The implementation of k_sys_fatal_error_handler(), which in most cases will be overridden by the default implementation provided here, completely determines what happens. It could do anything it likes.

andrewboie replied:
> How is the optional behavior configured for spinning or powering off?

It's literally the k_sys_fatal_error_handler() function body.
Note that is a weak function, which can be overridden by applications to implement their own policy.

dbkinder replied:
Ah.  So how about something like this:
```
In the event of an unhandled CPU exception, the architecture
code must call into :c:func:`z_fatal_error`.  This function dumps
out architecture-agnostic information and makes a policy
decision on what to do next by invoking ``k_sys_fatal_error()``.
This function can be overridden to implement application-specific
policies that could include locking interrupts and spinning forever
(the default implementation) or even powering off the
system (if supported).
```

andrewboie replied:
Thanks, will do

andrewboie replied:
done

dbkinder commented on doc/guides/porting/arch.rst:
```diff
@@ -441,13 +441,19 @@ However, a real implementation is strongly recommended.
+implement :c:func:`z_arch_system_halt` to power off the system. The default
+implementation locks interrupts and spins forever.
+
+In the event of an unhandled CPU exception, the architecture
```
Hmmm. My intention was to replace the previous paragraph with this one, deleting lines 444-448

andrewboie replied:
derp, fixed

dbkinder reviewed (approved):
Doc changes LGTM, thanks!

nordic-krch commented on kernel/fatal.c:
```diff
@@ -0,0 +1,134 @@
+/* LCOV_EXCL_STOP */
+
+#if defined(CONFIG_PRINTK) || defined(CONFIG_LOG)
+void z_fatal_print(const char *fmt, ...)
```
i would convert it to macro like:
```
#define z_fatal_print(...) \
    if (IS_ENABLED(CONFIG_LOG)) LOG_ERR(__VA_ARGS__); \
    else if(IS_ENABLED(CONFIG_PRINTK) printk(__VA_ARGS__);
```

then you can remove `#if defined(CONFIG_PRINTK) || defined(CONFIG_LOG)` from most part of code here (maybe even all).

andrewboie replied:
Done this partially, but I'm encountering an #include hell if I try to put that macro directly in kernel/fatal.h because kernel.h needs it for z_except_reason(). I could move some stuff around and remove fatal.h from kernel_includes.h, but this would break anyone using k_oops()/k_panic() without changing their code.

For now, I'd like to leave as a C function, but otherwise I'm incorporating the substance of your comment.

nordic-krch commented on kernel/fatal.c:
```diff
@@ -0,0 +1,134 @@
+				      const z_arch_esf_t *esf)
+{
+	ARG_UNUSED(esf);
+
```
if this is the default place for unrecoverable error, i would put `LOG_PANIC` here

andrewboie replied:
Done

nordic-krch commented on kernel/fatal.c:
```diff
@@ -0,0 +1,134 @@
+{
+#if defined(CONFIG_PRINTK) || defined(CONFIG_LOG)
+	struct k_thread *thread = k_current_get();
+	const char *thread_name = k_thread_name_get(thread);
```
for code purity i would create local `thread_name_get` function and use it directly in the log call:
`z_fatal_print("Current thread: %p (%s)", thread, thread_name_get(thread));`

andrewboie replied:
Done

nordic-krch commented on kernel/fatal.c:
```diff
@@ -0,0 +1,134 @@
+		thread_name = "unknown";
+	}
+
+	switch (reason) {
```
imo, it would be cleaner to extract it to function `reason_to_str()` and do something like `z_fatal_print(">>> ZEPHYR FATAL ERROR: %s (reason %d)", reason_to_str(reason), reason);`. it will be also less rom used since there will be only one call to printing instead of multiple.

andrewboie replied:
Done

nordic-krch commented on kernel/fatal.c:
This should probably be wrapped in `if (!k_is_in_isr())`. I'm not sure what `k_current_get` returns when called from interrupt context. Is it interrupted thread or NULL?. If it returns last active thread then it's not good to abort it due to fault in an interrupt and I see that `z_impl_k_thread_abort` is vulnerable to `thread == NULL`.

andrewboie replied:
@nordic-krch see the comment a just a few lines above, k_is_in_isr() isn't working right. I filed #17656 , meanwhile I would like to leave as-is.

> I see that z_impl_k_thread_abort is vulnerable to thread == NULL.

_kernel->current isn't ever NULL in this scenario

nordic-krch replied:
> kernel->current isn't ever NULL in this scenario

Is it because we assume that in interrupt it will never be treated as recoverable? Or is it because `k_current_get` never returns NULL?

What about adding `fixme` comment here as well then?

andrewboie replied:
>  k_current_get never returns NULL?

This.

> What about adding fixme comment here as well then?

I made this an assertion if CONFIG_TEST is disabled, and put a fixme note.

nordic-krch commented on kernel/fatal.c:
what about changing it to macro like:
```
#define z_fatal_print(fmt,...) \
    if (IS_ENABLED(CONFIG_LOG)) LOG_ERR(fmt,__VA_ARGS__); \
    else if (IS_ENABLED(CONFIG_PRINTK)) printk("FATAL: " fmt "\n", __VA_ARGS__) 
```
this macro reduces number of `#if defined(CONFIG_LOG) || defined(CONFIG_PRINTK)` as compiler will take care of what is used and what unused.

note that there is no `LOG_PANIC` there as it should be called explicitly at the place where decision about recovering is made.

andrewboie replied:
I already told you. I can't put this in the header. It results in an #include dependency loop. Fixing that would make this patch even more invasive than it already is. We need to move on.

andrewboie replied:
I removed the LOG_PANIC() call, that wasn't supposed to be there, it gets called in k_sys_fatal_error_handler() only.

andrewboie replied:
Also please note: the ifdefs are still necessary. gc-sections doesn't work the way you might expect and the string literals are not discarded even if the functions are unused. See https://stackoverflow.com/questions/54996229/is-ffunction-sections-fdata-sections-and-gc-sections-not-working

nordic-krch replied:
hm, that i didn't know. However, I've did simple test in hello world and it seems like they are removed.

> I already told you

Sorry, i missed that. Yes, macro may result in include looping :/ hopefully we can fix it later.

nordic-krch replied:
created #17841 which should cut the loop.

nordic-krch commented on kernel/fatal.c:
source is not set, it will be printed with random source prefix. I think it's better to switch to the macro as suggested in the comment above.

andrewboie replied:
fixed

vonhust commented:
I like this improvement, make code better.  will do further review

andyross reviewed (approved):
I can't find anything more to complain about.  This looks great to me.  We can continue to refine where needed, but this is the right framework.
